File: 2170-count-number-of-maximum-bitwise-or-subsets/2170-count-number-of-maximum-bitwise-or-subsets.cpp
class Solution {
public:
    int countMaxOrSubsets(vector<int>& nums) {
        int res = 0;
        int maxOr = 0;
        for (int i = 0; i < nums.size(); i++) {
            maxOr |= nums[i];
        }
        solve(nums, 0, maxOr, 0, res);
        return res;
    }

private:
    void solve(vector<int>& nums, int idx, int& maxOr, int cur, int& count) {
        if (idx == nums.size()) {
            return;
        }
        if ((cur | nums[idx]) == maxOr) {
            count++;
        }
        solve(nums, idx + 1, maxOr, cur | nums[idx], count);
        solve(nums, idx + 1, maxOr, cur, count);
    }
};

File: 2170-count-number-of-maximum-bitwise-or-subsets/2170-count-number-of-maximum-bitwise-or-subsets.py
class Solution:
    def solve(self, nums: List[int], idx: int, cur: int, maxOr: int) -> int:
        if idx == len(nums):
            return 1 if cur == maxOr else 0
        count = 0
        count += self.solve(nums, idx + 1, cur | nums[idx], maxOr)
        count += self.solve(nums, idx + 1, cur, maxOr)
        return count


    def countMaxOrSubsets(self, nums: List[int]) -> int:
        maxOr = 0
        for num in nums:
            maxOr |= num
        res = self.solve(nums, 0, 0, maxOr)
        return res

        